src/thunderbots/software/gui/widgets/main_widget.cpp

```diff
@@ -0,0 +1,464 @@
+    update();
+}
+
+MainWidget::~MainWidget()
+{
+    delete main_widget;
+    delete scene;
+    delete glWidget;
+}
+
+void MainWidget::draw(WorldDrawFunction world_draw_function,
+                      AIDrawFunction ai_draw_function)
+{
+    scene->clear();
+    world_draw_function.execute(scene);
+    ai_draw_function.execute(scene);
+    if (first_draw_call)
+    {
+        main_widget->ai_visualization_graphics_view->fitInView(scene->sceneRect(),
+                                                               Qt::KeepAspectRatio);
+        first_draw_call = false;
+    }
+}
+
+void MainWidget::updatePlayInfo(const PlayInfo& play_info)
+{
+    QString play_type_string =
+        QString("Play Type: %1\n").arg(QString::fromStdString(play_info.play_type));
+    QString play_name_string =
+        QString("Play Name: %1\n").arg(QString::fromStdString(play_info.play_name));
+    QString tactics_string = QString("Tactics:\n");
+    for (const auto& tactic_string : play_info.robot_tactic_assignment)
+    {
+        tactics_string.append(QString::fromStdString(tactic_string)).append("\n");
+    }
+
+    QString play_info_string =
+        QString("%1\n%2\n%3").arg(play_type_string, play_name_string, tactics_string);
+
+    main_widget->play_and_tactic_info_text_edit->setText(play_info_string);
+}
+
+void MainWidget::updateRobotStatus(const RobotStatus& robot_status)
+{
+    main_widget->robot_status_table_widget->updateRobotStatus(robot_status);
+}
+
+void MainWidget::setupSceneView(QGraphicsView* view, QGraphicsScene* scene,
+                                QOpenGLWidget* gl_widget)
+{
+    view->setScene(scene);
+    view->setDragMode(QGraphicsView::ScrollHandDrag);
+    view->setBackgroundBrush(QBrush(Qt::darkGreen, Qt::SolidPattern));
+
+    // Performance optimizations
+    // https://stackoverflow.com/questions/43826317/how-to-optimize-qgraphicsviews-performance
+    view->setInteractive(false);
+    view->setOptimizationFlag(QGraphicsView::DontAdjustForAntialiasing);
+    view->setCacheMode(QGraphicsView::CacheBackground);
+    view->setViewportUpdateMode(QGraphicsView::SmartViewportUpdate);
+    // Using an OpenGL widget with the view should help make use of the graphics card
+    // rather than doing CPU drawing, which should take some load off the CPU and make
+    // things faster
+    view->setViewport(gl_widget);
+
+    // Invert the y-coordinates of the view.
+    // We do this because Qt's default coordinate system for drawing is:
+    // * positive x = "right"
+    // * positive y = "down"
+    // Our coordinate system defines positive y as being "up", so we invert the coordinate
+    // system so all our draw calls can follow our coordinate convention. This also fixes
+    // the orientation convention, so "positive" rotation is counterclockwise in the view
+    // (rotating from +x, to +y, to -x, to -y)
+    QTransform view_transform(1, 0, 0, -1, 0, 0);
+    view->setTransform(view_transform);
+
+    view->update();
+}
+
```

Comment:
Can we split out some stuff in this file into seperate `ai_control`, `parameters`, `robot_status`, and `world_view` files?